### PR TITLE
[WIP] Add early stopping to the trainer

### LINKED ISSUE #4894: 🚀 Add early stopping to the trainer
# 🚀 Feature request

The trainer ([pt](https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py), [tf](https://github.com/huggingface/transformers/blob/master/src/transformers/trainer_tf.py)) is an easy access point for users who rather not spend too much time building their own trainer class but prefer an out-of-the-box solution. Even though `transformers` was never meant to be a fully fletched training library, it might please users to add an additional feature: early stopping. 

## Motivation

Early stopping ensures that the trainer does not needlessly keep training when the loss does not improve. This saves time, money, and let's not forget the trees. 😉 Performance-wise this should not lead to different results.

## Your contribution

At the moment I cannot work on this, but here are my thoughts:

- a training argument should be added ([pt](https://github.com/huggingface/transformers/blob/master/src/transformers/training_args.py), [tf](https://github.com/huggingface/transformers/blob/master/src/transformers/trainer_tf.py)). This would only work when `evaluate_during_training` is enabled.
- for PyTorch: at every evaluation step, an early stopper (can be a separate class even) checks if the loss has improved in the last n steps. Potentially with a minimal threshold that the loss should have improved. If not, the trainer should stop
- for Tensorflow: I don't have experience with TF myself, but I assume one could use [`tf.keras.callbacks.EarlyStopping`](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/EarlyStopping).

### PR DESCRIPTION
**Pull request by primaprashant**
closes #4894

### CODE REVIEW

**Comment by BramVanroy** [comment]
Well that was quick. Awesome! Let me know if you need some help (for the pytorch part) or code review.

**Comment by misrasaurabh1** [comment]
Looks like a duplicate of https://github.com/huggingface/transformers/pull/4186

**Comment by BramVanroy** [comment]
> Looks like a duplicate of #4186

You are absolutely right. Closing this in favour of https://github.com/huggingface/transformers/pull/4186
